### PR TITLE
Add tests for image checks

### PR DESCRIPTION
**Pull request by JaurbanRH**

### CODE REVIEW

File: testsuite/tests/system/test_system_image.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def test_system_image(images, openshift):
+    """
+    Test:
+        - load expected image from yaml file
+        - load deployed image from amp-system deployment
+        - assert that expected and deployed images are the same
+    """
+    expected_image = images.get("3scale")
```
Comment (mdujava) [inline testsuite/tests/system/test_system_image.py:14]: ambiguous name, I would add _system_app to it

File: testsuite/tests/system/test_system_image.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    if not expected_image:
+        warn_and_skip("Image of 3scale system not provided.")
+    lookup = openshift().do_action("get", ["is/amp-system", "-o", "yaml"])
```
Comment (mdujava) [inline testsuite/tests/system/test_system_image.py:17]: is there a reason why this is checking `is` and not the `pod` itself?

File: config/settings.yaml.tpl [hunk 3]
```diff
@@ -148,3 +148,8 @@ development:
     url: "" # URL for hyperfoil controller
   cfssl:
     binary: "cfssl" # Path to the cfssl binary
+  images:
+    apicast: "" # Image of deployed apicast
+    apicast_operator: "" # Image of apicast operator
+    3scale: "" # Image of deployed 3scale system
```
Comment (mdujava) [inline config/settings.yaml.tpl:154]: I would specify that it's system-app

**Review by mdujava — CHANGES_REQUESTED** [review]

File: testsuite/tests/images/test_images_check.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        warn_and_skip(f"Image of 3scale {image} not provided.")
+
+    openshift = openshift()
+
+    lookup = openshift.do_action("get", [f"is/{image_stream}", "-o", "yaml"])
+    lookup = yaml.safe_load(lookup.out())
+    digest = lookup["spec"]["tags"][-1]["from"]["name"].split(":")[-1]
+    assert digest == expected_image["digest"]
+
+    lookup = openshift.do_action("get", [f"dc/{deployment_config}", "-o", "yaml"])
```
Comment (pehala) [inline testsuite/tests/images/test_images_check.py:31]:
```
lookup = openshift.do_action("get", [f"dc/{deployment_config}", "-o", "yaml"], parse_output=True)
digest = lookup.model.spec.template.spec.containers[-1].image.split(":")[-1)
```

File: config/settings.yaml.tpl [hunk 5]
```diff
@@ -148,3 +148,19 @@ development:
     url: "" # URL for hyperfoil controller
   cfssl:
     binary: "cfssl" # Path to the cfssl binary
+  images:
+    apicast: # Digest of deployed apicast
+    apicast_operator: # Digest of apicast operator
+    3scale_backend:
+      digest: # Digest of deployed 3scale backend
```
Comment (pehala) [inline config/settings.yaml.tpl:155]: either:
`      digest: # Multi-arch manifest digest`
or
`      manifest_digest: # Multi-arch manifest digest`

**Comment by mganisin** [comment]
`oc cluster-info dump | jq '.items[].metadata.labels["kubernetes.io/arch"]'`

File: testsuite/tests/conftest.py [hunk 6]
```diff
@@ -934,3 +934,20 @@ def _prepare_prometheus_endpoint(routes):
...
+def images(testconfig):
+    """Load images file"""
+    images = testconfig.get("images")
+    if not images:
+        warn_and_skip("Images not provided.")
+    return images
+
+
+@pytest.fixture(scope="session")
+def cluster_arch(openshift):
```
Comment (mganisin) [inline testsuite/tests/conftest.py:949]: This should be `openshift.arch` not a fixture

File: testsuite/tests/images/test_images_check.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+                          ("apicast", "amp-apicast", ["apicast-staging", "apicast-production"])])
+@pytest.mark.nopersistence
+def test_image_digest(images, openshift, image, image_stream, deployment_configs, cluster_arch):
+    """
+    Test:
+        - load expected images from settings
+        - assert that expected image and image in image stream are the same
+        - assert that expected image and image in deployment config are the same
+    """
+    expected_image = images.get(f"{image}")
```
Comment (pehala) [inline testsuite/tests/images/test_images_check.py:22]: No need for string formatting

**Review by pehala — DISMISSED** [review]

**Review by mganisin — APPROVED** [review]

File: testsuite/tests/images/test_images_check.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    """return: Self-managed Apicast operator"""
+    namespace = testconfig["threescale"]["gateway"]["OperatorApicast"]['openshift']['project_name']
+    openshift = openshift(project=namespace)
+    return openshift.get_apicast_operator().object()
+
+
+@pytest.fixture(scope="session")
+def threescale_operator(testconfig, openshift):
+    """return: 3scale operator"""
+    namespace = testconfig["openshift"]["projects"]["threescale"]['name']
```
Comment (pehala) [inline testsuite/tests/images/test_images_check.py:72]: Why do you need the namespace, shouldn't just openshift() do the job?

File: testsuite/tests/images/test_images_check.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    digest = lookup.model.spec.template.spec.containers[0].image.split(":")[-1]
+    assert digest == expected_image["manifest_digest"]
+
+
+@pytest.fixture(scope="session")
+def apicast_operator(testconfig, openshift):
+    """return: Self-managed Apicast operator"""
+    namespace = testconfig["threescale"]["gateway"]["OperatorApicast"]['openshift']['project_name']
+    openshift = openshift(project=namespace)
+    return openshift.get_apicast_operator().object()
```
Comment (pehala) [inline testsuite/tests/images/test_images_check.py:66]: `staging_gateway.openshift.get_apicast_operator().object()`, you already create SelfManagedApicast

File: testsuite/tests/images/test_images_check.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        lookup = openshift.do_action("get", [f"dc/{deployment_config}", "-o", "yaml"], parse_output=True)
+        for container in lookup.model.spec.template.spec.containers:
+            digest = container.image.split(":")[-1]
+            assert digest == expected_image["resolved_images"].get(openshift.arch)
+
+
+@pytest.fixture(scope="module")
+def staging_gateway(request, testconfig):
+    """Deploy self-managed template based apicast gateway."""
+    gw = gateway(kind=SelfManagedApicast, staging=True, name=blame(request, "gw"))
```
Comment (pehala) [inline testsuite/tests/images/test_images_check.py:40]: I would use just OperatorApicast, I dont see any reason to run it on Template

**Review by pehala — APPROVED** [review]

**Review by dhlavac — APPROVED** [review]

**Pull request merged**